res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,772 @@
+            } else if (value === 0x7F) {
+                engine.setParameter("[Channel1]", "waveform_zoom_up", 1);
+
+                // see above comment
+                if (!MixtrackProFX.waveformsSynced) {
+                    engine.setParameter("[Channel2]", "waveform_zoom_up", 1);
+                }
+            }
+        }
+    });
+
+    this.knobButton = new components.Button({
+        group: "[Library]",
+        inKey: "MoveFocusForward"
+    });
+
+    this.buttonShift = new components.Button({
+        group: "[Library]",
+        inKey: "GoToItem"
+    });
+
+    this.setBeatgrid = new components.Button({
+        group: "[Channel1]",
+        key: "beats_translate_curpos",
+        midi: [0x88, 0x01],
+        off: 0x01,
+        shift: function() {
+            this.group = "[Channel2]";
+        },
+        unshift: function() {
+            this.group = "[Channel1]";
+        }
+    });
+};
+
+MixtrackProFX.Browse.prototype = new components.ComponentContainer();
+
+MixtrackProFX.HeadGain = function() {
+    components.Pot.call(this);
+};
+
+MixtrackProFX.HeadGain.prototype = new components.Pot({
+    group: "[Master]",
+    inKey: "headGain"
+});
+
+MixtrackProFX.vuCallback = function(value, group) {
+    var level = value * 90;
+
+    if (engine.getValue("[Channel1]", "pfl") || engine.getValue("[Channel2]", "pfl")) {
+        if (group === "[Channel1]") {
+            midi.sendShortMsg(0xB0, 0x1F, level);
+        } else if (group === "[Channel2]") {
+            midi.sendShortMsg(0xB1, 0x1F, level);
+        }
+    } else if (group === "[Channel1]") {
+        midi.sendShortMsg(0xB0, 0x1F, level);
+    } else if (group === "[Channel2]") {
+        midi.sendShortMsg(0xB1, 0x1F, level);
+    }
+};
+
+MixtrackProFX.scratchToggle = function(channel) {
+    MixtrackProFX.scratchModeEnabled[channel] = !MixtrackProFX.scratchModeEnabled[channel];
+    midi.sendShortMsg(0x90 | channel, 0x07, MixtrackProFX.scratchModeEnabled[channel] ? 0x7F : 0x01);
+};
+
+MixtrackProFX.wheelTouch = function(channel, control, value) {
+    var deckNumber = channel + 1;
+
+    if (!MixtrackProFX.shifted && MixtrackProFX.scratchModeEnabled[channel] && value === 0x7F) {
+        // touch start
+
+        engine.scratchEnable(deckNumber, MixtrackProFX.jogScratchSensitivity, 33+1/3, MixtrackProFX.jogScratchAlpha, MixtrackProFX.jogScratchBeta, true);
+    } else if (value === 0) {
+        // touch end
+        engine.scratchDisable(deckNumber, true);
+    }
+};
+
+MixtrackProFX.wheelTurn = function(channel, control, value, status, group) {
+    var deckNumber = channel + 1;
+
+    var newValue = value;
+
+    if (value >= 64) {
+        // correct the value if going backwards
+        newValue -= 128;
+    }
+
+    if (MixtrackProFX.shifted) {
+        // seek
+        var oldPos = engine.getValue(group, "playposition");
+
+        if (oldPos <= 0) {
+            oldPos = 0;
+        }
+
+        if (oldPos >= 1) {
+            oldPos = 1;
+        }
```

Comment:
I don't see any practical benefit in clamping here

Reply:
It doesn't let you scroll backwards into infinity. Should that be allowed?

Reply:
Well, we're allowing it in mixxx. There are legitimate usecases for navigating into parts before the track starts.